src/modules/WithParamsAndPublicKeys.js

```diff
@@ -0,0 +1,357 @@
+    const tx = this.createNewPublicKeyTx(addPublicKey, did, keyPair, signature);
+    return this.signAndSend(tx, waitForFinalization, params);
+  }
+
+  async removeParams(did, counter, keyPair = undefined, signature = undefined, waitForFinalization = true, params = {}) {
+    const tx = this.removeParamsTx(did, counter, keyPair, signature);
+    return this.signAndSend(tx, waitForFinalization, params);
+  }
+
+  async removePublicKey(did, counter, keyPair = undefined, signature = undefined, waitForFinalization = true, params = {}) {
+    const tx = this.removePublicKeyTx(did, counter, keyPair, signature);
+    return this.signAndSend(tx, waitForFinalization, params);
+  }
+
+  // eslint-disable-next-line no-unused-vars
+  async queryParamsFromChain(hexDid, counter) {
+    throw new Error('Extending class should implement queryParamsFromChain');
+  }
+
+  // eslint-disable-next-line no-unused-vars
+  async queryPublicKeyFromChain(hexDid, counter) {
+    throw new Error('Extending class should implement queryPublicKeyFromChain');
+  }
+
+  async getParams(did, counter) {
+    const hexId = getHexIdentifierFromDID(did);
+    return this.getParamsByHexDid(hexId, counter);
+  }
+
+  /**
+   *
+   * @param did
+   * @param counter
+   * @param withParams - If true, return the params referenced by the public key. It will throw an error if params_ref is null
+   * or params were not found on chain which can happen if they were deleted after this public key was added.
+   * @returns {Promise<{bytes: string}|null>}
+   */
+  async getPublicKey(did, counter, withParams = false) {
+    const hexId = getHexIdentifierFromDID(did);
+    return this.getPublicKeyByHexDid(hexId, counter, withParams);
+  }
+
+  async getParamsByHexDid(hexDid, counter) {
+    const resp = await this.queryParamsFromChain(hexDid, counter);
+    if (resp.isSome) {
+      return this.createParamsObjFromChainResponse(resp.unwrap());
+    }
+    return null;
+  }
+
+  async getPublicKeyByHexDid(hexDid, counter, withParams = false) {
+    const resp = await this.queryPublicKeyFromChain(hexDid, counter);
+    if (resp.isSome) {
+      const pkObj = this.createPublicKeyObjFromChainResponse(resp.unwrap());
+      if (withParams) {
+        if (pkObj.params_ref === null) {
+          throw new Error('No reference to parameters for the public key');
+        } else {
+          const params = await this.getParamsByHexDid(pkObj.params_ref[0], pkObj.params_ref[1]);
+          if (params === null) {
+            throw new Error(`Parameters with reference (${pkObj.params_ref[0]}, ${pkObj.params_ref[1]}) not found on chain`);
+          }
+          pkObj.params = params;
+        }
+      }
+      return pkObj;
+    }
+    return null;
+  }
+
+  /**
+   * Get all params written by a DID
+   * @param did
+   * @returns {Promise<object[]>}
+   */
+  async getAllParamsByDid(did) {
+    const hexId = getHexIdentifierFromDID(did);
+
+    // TODO: Figure out why this doesn't work
```

Comment:
is this comment still relevant?

Reply:
Its relevant but not important now. I can do this optimization later